src/main/java/wingman/logic/toplevel/link/LinkFactoryBase.java

```diff
@@ -0,0 +1,164 @@
+     */
+    private final Lazy<ReadOnlyItemManager<T>> targetManagerLazy;
+
+    /**
+     * Creates a new link factory.
+     *
+     * @param sourceManagerLazy the source manager.
+     * @param targetManagerLazy the target manager.
+     */
+    public LinkFactoryBase(
+            Lazy<ReadOnlyItemManager<S>> sourceManagerLazy,
+            Lazy<ReadOnlyItemManager<T>> targetManagerLazy
+    ) {
+        this.sourceManagerLazy = sourceManagerLazy;
+        this.targetManagerLazy = targetManagerLazy;
+    }
+
+    /**
+     * Gets the error message shown for the command format.
+     *
+     * @return the error message shown for the command format.
+     */
+    protected abstract String getCommandFormatHint();
+
+    /**
+     * Gets the item or throws the corresponding exception.
+     *
+     * @param prefix     the prefix of item.
+     * @param idOptional the id of the item.
+     * @param manager    the manager of the item.
+     * @param <I>        the prefix of item.
+     * @return the item.
+     */
+    protected <I extends Item> I getItemOrThrow(
+            String prefix,
+            Optional<String> idOptional,
+            ReadOnlyItemManager<I> manager
+    ) throws ParseException {
+        if (idOptional.isEmpty()) {
+            throw ParseException.formatted(
+                    "%s: index is missing for %s.\nCorrect Format: %s",
+                    getCommandWord(),
+                    prefix,
+                    getCommandFormatHint()
+            );
+        }
+        int flightId;
+        try {
+            flightId = Integer.parseInt(idOptional.get());
+        } catch (NumberFormatException e) {
+            throw NumberParseException.formatted(
+                    "%s: cannot parse %s to a number for prefix %s: %s"
+                            + ".\nCorrect Format: %s",
+                    getCommandWord(),
+                    idOptional.get(),
+                    prefix,
+                    e.getMessage(),
+                    getCommandFormatHint()
+            );
+        }
+        boolean isIndexValid = flightId >= 1 && flightId <= manager.size();
+        if (!isIndexValid) {
+            throw NumberParseException.formatted(
+                    "%s: index %d is out of bounds for prefix %s, "
+                            + "should be between 1 and %d. \nCorrect Format: %s",
+                    getCommandWord(),
+                    flightId,
+                    prefix,
+                    manager.size(),
+                    getCommandFormatHint()
+            );
+        }
+        Optional<I> flightOptional = manager.getItemOptional(flightId - 1);
+        if (flightOptional.isEmpty()) {
+            throw NumberParseException.formatted(
+                    "%s: with index %d is not found."
+                            + "\nCorrect Format: %s",
+                    getCommandWord(),
+                    flightId,
+                    getCommandFormatHint()
+            );
+        }
```

Comment:
A little bit long, but this is not common in the entire code base. Then it should be good.